### PR TITLE
Fix visitPHINode for PatchBufferOp

### PR DESCRIPTION
**Pull request by LLJJDD**
When visiting a phi node in PatchBufferOp, an incoming value may be unvisited and result in null when getting it from m_replacementMap. A shader to reproduce this issue is included in the commit.

This commit fixes the issue by calling the visit method when a value cannot be gotten from m_replacementMap, so there will be a valid value in the map.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 83f5ce10284d3a29b3128ac8b42ccd2800be38d0
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev 3909a9fa55548c66e069a04ff79c96c8a7260851</li>
<li>SPVGEN: dev 8dc855026f2502ab3f45dadaf0bb802a57d6ad60</li>
<li>XGL: dev f4a992dd7e556ed5e7e2ffa2c830f1cd79bd4596</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c2ea9a11546d054e3918c44ac3b3ad930686cd79</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/130900)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 22784/43839 (52.0%)</li>
<li>Failed: 0/43839 (0.0%)</li>
<li>Skipped: 21055/43839 (48.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 22783/43843 (52.0%)</li>
<li>Failed: 0/43843 (0.0%)</li>
<li>Skipped: 21060/43843 (48.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 21609/43218 (50.0%)</li>
<li>Failed: 0/43218 (0.0%)</li>
<li>Skipped: 21609/43218 (50.0%)</li>
</ul></details>


</ul></details>

**Comment by nhaehnle** [comment]
An alternative approach would be to visit the phi nodes in two phases. When an incoming value isn't yet in `m_replacementMap`, leave the incoming value as undef and record it in a global list. After the whole function has been visited, go through the list and fixup the remaining incoming values of phi nodes.

Or insert incomplete phi nodes unconditionally, add them to the `m_replacementMap`, and then do all the phi nodes in one go after the main visit has completed.

**Comment by LLJJDD** [comment]
Thanks for the suggestions :)

I'm unclear about just one thing:

> Your patch fixes this if the def in this case is also a phi node, but I think it might fail if the def is some other instruction.

- I assume the `def` here is the `incomings[blockIndex]` in the code, so it is actually a GEP inst in my case. By calling `visit(*inst)`, I think it should be able to handle any possible instruction.

**Comment by LLJJDD** [comment]
All right, now I'm clear about the issue as @ruiling has talked to me offline - during `visit(*inst)`, the src of it may also be unvisited, that is the problem.

Then I prefer your second approach since it is more straight forward, I will work on it.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 2a201754ec3aaf504486e78cfcf90c2d25bfd966
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev 3909a9fa55548c66e069a04ff79c96c8a7260851</li>
<li>SPVGEN: dev 8dc855026f2502ab3f45dadaf0bb802a57d6ad60</li>
<li>XGL: dev f4a992dd7e556ed5e7e2ffa2c830f1cd79bd4596</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev ea979e90c306a6ddec5db687e9c8d33b5289a23c</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/131499)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 22784/43839 (52.0%)</li>
<li>Failed: 0/43839 (0.0%)</li>
<li>Skipped: 21055/43839 (48.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 22783/43843 (52.0%)</li>
<li>Failed: 0/43843 (0.0%)</li>
<li>Skipped: 21060/43843 (48.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 21609/43817 (49.3%)</li>
<li>Failed: 0/43817 (0.0%)</li>
<li>Skipped: 22208/43817 (50.7%)</li>
</ul></details>


</ul></details>

**Comment by LLJJDD** [comment]
@nhaehnle please kindly have a look at the updated commit. Thanks!

File: llpc/test/shaderdb/TestPatchBufferOp.comp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+
+// BEGIN_SHADERTEST
+/*
+; RUN: amdllpc -spvgen-dir=%spvgendir% -v %gfxip %s | FileCheck -check-prefix=SHADERTEST %s
+; SHADERTEST-LABEL: {{^// LLPC}} SPIRV-to-LLVM translation results
+; SHADERTEST: AMDLLPC SUCCESS
+*/
+// END_SHADERTEST
```
Comment (kuhar) [inline llpc/test/shaderdb/TestPatchBufferOp.comp:44]: Could you move it to the top of the file? I find it easier to figure out what a test is doing or why it's failing when RUN lines are towards the top.
Summarizing what a test is supposed to exercise in plain English is also a good practice.

File: lgc/patch/PatchBufferOp.cpp [hunk 2]
```diff
@@ -1664,6 +1672,20 @@ Instruction *PatchBufferOp::makeLoop(Value *const loopStart, Value *const loopEn
   return loopCounter;
 }
 
+// =====================================================================================================================
+// Fix incomplete phi incoming value
+void PatchBufferOp::fixIncompletePhi() {
+  for (auto phiIt = m_incompletePhi.begin(); phiIt != m_incompletePhi.end(); ++phiIt) {
```
Comment (kuhar) [inline lgc/patch/PatchBufferOp.cpp:1678]: nit: use a range-based for loop

File: lgc/patch/PatchBufferOp.cpp [hunk 3]
```diff
@@ -733,7 +736,12 @@ void PatchBufferOp::visitPHINode(PHINode &phiNode) {
       const int blockIndex = phiNode.getBasicBlockIndex(block);
       assert(blockIndex >= 0);
 
-      Value *const incomingBufferDesc = m_replacementMap[incomings[blockIndex]].first;
+      Value *incomingBufferDesc = m_replacementMap[incomings[blockIndex]].first;
+
+      if (!incomingBufferDesc) {
+        incomingBufferDesc = UndefValue::get(newPhiNode->getType());
+        m_incompletePhi[std::make_pair(newPhiNode, block)] = incomings[blockIndex];
```
Comment (kuhar) [inline lgc/patch/PatchBufferOp.cpp:743]: nit: you can use `{newPhiNode, block}` instead of make_pair

File: lgc/patch/PatchBufferOp.cpp [hunk 4]
```diff
@@ -733,7 +736,12 @@ void PatchBufferOp::visitPHINode(PHINode &phiNode) {
...
+      }
```
Comment (ruiling) [inline lgc/patch/PatchBufferOp.cpp:744]: Could you also change the 'offset' part in the following to use this way? adding an extra visit() may not be robust enough. I mean the code around line 770.

Comment (LLJJDD) [reply lgc/patch/PatchBufferOp.cpp:746]: Sure. Done in latest commit.

File: llpc/test/shaderdb/TestPatchBufferOp.comp (hunk 1 above)
Comment (LLJJDD) [reply llpc/test/shaderdb/TestPatchBufferOp.comp]: Sure, all comments are addressed, thanks for them.

File: lgc/patch/PatchBufferOp.cpp [hunk 5]
```diff
@@ -1664,6 +1673,26 @@ Instruction *PatchBufferOp::makeLoop(Value *const loopStart, Value *const loopEn
   return loopCounter;
 }
 
+// =====================================================================================================================
+// Fix incomplete phi incoming value
+void PatchBufferOp::fixIncompletePhi() {
```
Comment (nhaehnle) [inline lgc/patch/PatchBufferOp.cpp:1678]: Please use the plural form:
```suggestion
// Fix incomplete phi incoming values
void PatchBufferOp::fixIncompletePhis() {
```

**Review by nhaehnle — COMMENTED** [review]
Thanks, this looks very nice now. One spelling nit inline, apart from that LGTM.

File: lgc/patch/PatchBufferOp.cpp (hunk 5 above)
Comment (LLJJDD) [reply lgc/patch/PatchBufferOp.cpp]: OK. I also change m_incompletePhi to m_incompletePhis.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 4c99b1c5ec702e554f791997d56f364137c0cae6
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev 3909a9fa55548c66e069a04ff79c96c8a7260851</li>
<li>SPVGEN: dev 8dc855026f2502ab3f45dadaf0bb802a57d6ad60</li>
<li>XGL: dev f4a992dd7e556ed5e7e2ffa2c830f1cd79bd4596</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev ccbc834ac4f92e63a9a003bd5a18cb9fcf3c6121</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/131507)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 22784/43840 (52.0%)</li>
<li>Failed: 0/43840 (0.0%)</li>
<li>Skipped: 21056/43840 (48.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 22783/43844 (52.0%)</li>
<li>Failed: 0/43844 (0.0%)</li>
<li>Skipped: 21061/43844 (48.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 21609/43823 (49.3%)</li>
<li>Failed: 0/43823 (0.0%)</li>
<li>Skipped: 22214/43823 (50.7%)</li>
</ul></details>


</ul></details>

**Comment by LLJJDD** [comment]
Hi @nhaehnle , if there is no other changes required, may you approve this PR? This is blocking the testing of one game. Thanks!

**Review by amdrexu — APPROVED** [review]
